msdk-io-mgf/src/main/java/io/github/msdk/io/mgf/MgfFileImportMethod.java

```diff
@@ -0,0 +1,173 @@
+import io.github.msdk.util.DataPointSorter.SortingProperty;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Hashtable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class MgfFileImportMethod implements MSDKMethod<List<MgfMsSpectrum>> {
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+  private boolean cancelled;
+  private List<MgfMsSpectrum> spectra;
+  private final @Nonnull File target;
+  private Hashtable<String, Pattern> patterns;
+
+  @Nullable
+  @Override
+  public List<MgfMsSpectrum> execute() throws MSDKException {
+    logger.info("Started MGF import from {} file", target);
+
+    try (final BufferedReader reader = new BufferedReader(new FileReader(target))) {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        switch (line) {
+          case "BEGIN IONS":
+            spectra.add(processSpectrum(reader));
+            break;
+        }
+      }
+      reader.close();
+    } catch (IOException e) {
+      throw new MSDKException(e);
+    }
+
+    return spectra;
+  }
+
+  private MgfMsSpectrum processSpectrum(BufferedReader reader) throws IOException, MSDKException {
+    String title = "";
+    int precursorCharge = 0;
+    double precursorMass = 0;
+    double mz[] = new double[16];
+    float intensity[] = new float[16];
+
+    int index = 0;
+    String line;
+    String groupped;
+    while (true) {
+      line = reader.readLine();
+      if (line == null || line.equals("END IONS") || cancelled) {
+        break;
+      }
+
+      // Not sure how to use cancelled variable
+
+      /*
+      * Code duplication (Matcher)
+      * TODO: Find a solution for it.
+      * */
+      try {
+        if (line.contains("PEPMASS")) {
+          Matcher m = patterns.get("PEPMASS").matcher(line);
+          m.find();
+          groupped = m.group();
+          precursorMass = Double.parseDouble(groupped);
+        } else if (line.contains("TITLE")) {
+          Matcher m = patterns.get("TITLE").matcher(line);
+          m.find();
+          title = m.group();
+        } else if (line.contains("CHARGE")) {
+          Matcher m = patterns.get("CHARGE").matcher(line);
+          m.find();
+          groupped = m.group();
+          String trimmed = groupped.substring(0, groupped.length() - 1);
+          precursorCharge = Integer.parseInt(trimmed);
+          if (groupped.charAt(groupped.length() - 1) == '-') {
+            precursorCharge *= -1;
+          }
+        } else {
+          String[] floats = line.split(" ");
+          double mzValue = Double.parseDouble(floats[0]);
+          float intensityValue = Float.parseFloat(floats[1]);
```

Comment:
what if the line does not contain the space (" ") character?

Reply:
I supppose the delimeter should be specified externally

Reply:
What I mean is you should make sure the length of the floats[] array is 2, before you start accessing floats[1]. If the array is smaller than 2 items, you will get an exception.